### DISCUSSION TITLE
MPC support?

### DISCUSSION THREAD

**Discussion by roangel** [pypose/pypose#140] [Q&A] [answered]
Hi guys, I see that you have added in the past some LTI, LTV and non-linear systems support (cartpole, floquet and neuralnet examples, respectively). I also see in the paper that you guys mention Differentiable-MPC, from Brandon Amos, and take it as a baseline. I have two questions:
- Does the current state of pypose already support MPC? I see it might not be the case because of this: https://github.com/pypose/pypose/issues/32
- Does pypose support hard constraints in the optimizations?

**Comment by wang-chen** [comment]
Hi @roangel Thank you for asking this, all the things you mentioned are currently under development, @aerogjy knows more about this.

**Comment by aerogjy** [comment, marked as answer]
Hi @roangel Thanks for the interest of our Differentiable MPC module! Currently, no. We will push the LQR module, which is the core part of D-MPC very soon, probably in a few weeks. Then, we will release the example we provided in the paper. 
The constraint optimization is on our high priority development list. We are on-going developing it right now. Please continue keeping an eye on that.

**Reply by roangel** [reply]
Thanks a lot for the replies and clarification! I need the constrained optimization part, so I'll stay tuned. Is there any collaboration chat where I can stay posted and maybe contribute in the future? Thanks!

**Reply by wang-chen** [reply]
@roangel If you are pretty sure that you will contribute to the repo, we strongly encourage you to [join our SLACK](https://join.slack.com/t/pypose/shared_invite/zt-17ezaltj8-eVHghcdJyF5nNuxQbcuMxw) for discussion and task scheduling, otherwise you may also post any concerns, questions, and suggestions here, we will keep you updated.

**Comment by wang-chen** [comment]
@roangel MPC has been released in [v0.5.0-beta](https://pypose.org/docs/main/generated/pypose.module.MPC/#pypose.module.MPC).
You may install via
```bash
pip install pypose==0.5.0b0
```
It is a beta release because we will do coding acceleration later but the API will not be changed, so you may feel free to use it.
The forward function args `u_lower`, `u_upper `, `du` will be supported in a future release.

**Reply by roangel** [reply]
Thanks a lot for this! I'll take a look!

**Reply by roangel** [reply]
Hi guys, any idea when the support for `du` will be released? Thanks!
